### DISCUSSION TITLE
[firebase_messaging] critical alerts on iOS?

### DISCUSSION THREAD

**Discussion by Finn0811** [firebase/flutterfire#5590] [Feature request]
Is it possible to send critical push notifications to iOS devices? 

[Apple documentation](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification#2990112) says it needs an 'critical' key in the sound dictionary but how do I have to do it with firebase messaging?

**Comment by Ehesp** [comment]
This should be possible, however it requires the critical permission to be set which is currently no implemented. 

Will leave open to track.

**Comment by davidpanic** [comment]
Has there been any work done on this?

**Reply by julianbraasch** [reply]
Yes you can now call ```await firebaseMessaging.requestPermission(criticalAlert: true);```. It still requires a special provisioning profile by Apple but that has nothing to do with Flutter or this plugin.

**Reply by davidpanic** [reply]
We already have the provisioning profile, thank you for letting us know about the `requestPermission` parameter, how would we go about sending that type of notification, e.g. what data should we set in the request to `https://fcm.googleapis.com/fcm/send`? We couldn't find anything about it in the documentation, I bet we're just missing something.

**Reply by julianbraasch** [reply]
Example message:
```java
Message message = Message
            .builder()
            .setToken(...)
            .setNotification(Notification
                .builder()
                .setTitle(...)
                .setBody(...)
                .build())
            .setApnsConfig(
                ApnsConfig
                    .builder()
                    .putAllHeaders(ImmutableMap
                        .<String, String>builder()
                        .put("apns-priority", "10")
                        .put("apns-expiration", String.valueOf(Instant.now().plus(1, ChronoUnit.DAYS).getEpochSecond()))
                        .build())
                    .setAps(
                        Aps
                            .builder()
                            .setSound(CriticalSound
                                .builder()
                                .setCritical(true)
                                .setVolume(1.0)
                                .setName("default")
                                .build())
                            .setBadge(badgeCount)
                            .build())
                    .build())
            .build();
```

**Reply by davidpanic** [reply]
I don't know what language that is or what you're using, but that wasn't my question. We're using the `https://fcm.googleapis.com/fcm/send` API to send the messages. Its input is a JSON formatted post body. The API reference is located [here](https://firebase.google.com/docs/cloud-messaging/http-server-ref). To my knowledge there is no mention of any parameter named "critical" or similar.
